# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/VideoFrame.h
# Commit: 1ec020b42625
# Full Hash: 1ec020b42625dba5da44b244027a97fc4ea8d623
# Author: Chun-Min Chang <chun.m.chang@gmail.com>
# Date: 2023-06-30 04:06:43
# Regressor Bug: 1831451
# File Overlap Count: 1
# Description:
#   Bug 1831451 - Allow constructing VideoFrame without pixel format r=padenot
#   
#   Depends on D171166
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D171518
# ==============================================================================

diff -r b49ae6bc356a -r 1ec020b42625 dom/media/webcodecs/VideoFrame.h
--- a/dom/media/webcodecs/VideoFrame.h	Thu Jun 29 22:33:13 2023 +0000
+++ b/dom/media/webcodecs/VideoFrame.h	Thu Jun 29 22:33:13 2023 +0000
@@ -55,13 +55,13 @@
 namespace mozilla::dom {
 
 struct VideoFrameData {
-  VideoFrameData(layers::Image* aImage, const VideoPixelFormat& aFormat,
+  VideoFrameData(layers::Image* aImage, const Maybe<VideoPixelFormat>& aFormat,
                  gfx::IntRect aVisibleRect, gfx::IntSize aDisplaySize,
                  Maybe<uint64_t> aDuration, int64_t aTimestamp,
                  const VideoColorSpaceInit& aColorSpace);
 
   const RefPtr<layers::Image> mImage;
-  const VideoPixelFormat mFormat;
+  const Maybe<VideoPixelFormat> mFormat;
   const gfx::IntRect mVisibleRect;
   const gfx::IntSize mDisplaySize;
   const Maybe<uint64_t> mDuration;
@@ -71,7 +71,7 @@
 
 struct VideoFrameSerializedData : VideoFrameData {
   VideoFrameSerializedData(layers::Image* aImage,
-                           const VideoPixelFormat& aFormat,
+                           const Maybe<VideoPixelFormat>& aFormat,
                            gfx::IntSize aCodedSize, gfx::IntRect aVisibleRect,
                            gfx::IntSize aDisplaySize, Maybe<uint64_t> aDuration,
                            int64_t aTimestamp,
@@ -89,7 +89,7 @@
 
  public:
   VideoFrame(nsIGlobalObject* aParent, const RefPtr<layers::Image>& aImage,
-             const VideoPixelFormat& aFormat, gfx::IntSize aCodedSize,
+             const Maybe<VideoPixelFormat>& aFormat, gfx::IntSize aCodedSize,
              gfx::IntRect aVisibleRect, gfx::IntSize aDisplaySize,
              const Maybe<uint64_t>& aDuration, int64_t aTimestamp,
              const VideoColorSpaceInit& aColorSpace);
@@ -216,15 +216,17 @@
   // A class representing the VideoFrame's data.
   class Resource final {
    public:
-    Resource(const RefPtr<layers::Image>& aImage, const Format& aFormat);
+    Resource(const RefPtr<layers::Image>& aImage, Maybe<Format>&& aFormat);
     Resource(const Resource& aOther);
     ~Resource() = default;
+    Maybe<VideoPixelFormat> TryPixelFormat() const;
     uint32_t Stride(const Format::Plane& aPlane) const;
     bool CopyTo(const Format::Plane& aPlane, const gfx::IntRect& aRect,
                 Span<uint8_t>&& aPlaneDest, size_t aDestinationStride) const;
 
     const RefPtr<layers::Image> mImage;
-    const Format mFormat;
+    // Nothing() if mImage is not in VideoPixelFormat
+    const Maybe<Format> mFormat;
   };
 
   nsCOMPtr<nsIGlobalObject> mParent;
